lib/AST/Decl.cpp

```diff
@@ -2246,6 +2246,12 @@ static Type mapSignatureFunctionType(ASTContext &ctx, Type type,
         type = objectType;
       }
     }
+    
+    // Replace an opaque archetype with a common dummy type. Functions and
+    // subscripts cannot overload differing only in opaque return types.
+    if (type->getAs<OpaqueTypeArchetypeType>()) {
+      type = BuiltinIntegerType::get(0x12345678u, ctx);
```

Comment:
Would erasing to the existential type be better? I'm worried this might leak through in diagnostics

Reply:
Maybe. That would mean that you couldn't overload:

```
func foo() -> any P
func foo() -> some P
```

I was also trying to prevent:

```
func foo() -> some P
func foo() -> some Q
```

since there would still be no way to disambiguate these.

Reply:
Couldn’t the latter be disambiguated by calling a generic function with a T: P or T: Q constraint with foo() as an argument?

IIRC the redeclaration checking uses a tuple of flags in addition to the erased overload type as the unique key. You could add a HasOpaqueResult flag and erase to an existential.